### ISSUE TITLE
Integrate ICON-CH1

### ISSUE THREAD

**Issue by terraputix** [open-meteo/open-meteo#1286]
Relevant Links: 
- https://github.com/MeteoSwiss/opendata-forecast-data/blob/main/README.md#2-numerical-weather-forecasting-model-data
- https://sys-data.int.bgdi.ch/browser/#/collections/ch.meteoschweiz.ogd-forecasting-icon-ch1

**Comment by patrick-zippenfenig** [comment]
The forecast data comes in the original ICON grid, which is more space-efficient. However, since ICON uses an unstructured triangular mesh, it's hard to work with in most tools. Like with the DWD ICON global data, I’d like to reproject it onto a regular lat-lon grid. Has anyone found the CDO reprojection weights needed to do this?

**Comment by terraputix** [comment]
I think the following python scripts demonstrate how they could be calculated, but I did not find any related output files:
- https://github.com/MeteoSwiss/meteodata-lab/blob/1a6b8d5091da715b78ea2ed0bf28265bf623e21f/jenkins/icontools/export_coeffs.py
- https://github.com/MeteoSwiss/meteodata-lab/blob/1a6b8d5091da715b78ea2ed0bf28265bf623e21f/src/meteodatalab/icon_grid.py

**Comment by patrick-zippenfenig** [comment]
Quickly created a plot. Seems like a good fit for a rotated lat-lon projection. I opened a ticket at MCH to ask for the weights

![Image](https://github.com/user-attachments/assets/77182cab-0d2d-4707-9bd5-3718a72d6064)

**Comment by guidocioni** [comment]
@patrick-zippenfenig isn't the grid already available inside the files? Your plot seems to suggest so, otherwise you wouldn't be able to put the raster on the right map. In this case you can create the weights yourself with CDO once you define a target grid.

**Comment by guidocioni** [comment]
Ok, I just tried to download some files of ICON-CH1 (why a REST API to download GRIB files MeteoSwiss...why?!?!?) and they don't have the `clon,clat` infos insitde. But I found the grid of ICON-CH1 (see attachment): you can use that to regrid with CDO.
Or, yes, generate the weights once and then remap so you save some time. 

BTW the `icontools` repo has some RBF methods to remap which are much faster (and precise) than the CDO ones. I know DWD is not keen on redistribute that source code (I have access anyway) but if you value speed it may be a good idea to implement a RBF interpolation in Swift if you need to process these data every day. 


[horizontal_constants.nc.zip](https://github.com/user-attachments/files/20271785/horizontal_constants.nc.zip)

**Comment by guidocioni** [comment]
Hey @patrick-zippenfenig, if you don't mind I'd like to clarify something.

The weights are useful only to speed up remapping with CDO, however you still need a target grid. For this reason, once you have a target grid, you could generate the weights yourself. I made some tests in the past and, for small grids as the one of ICON-CH1, using pre-generated weights does not save so much time. Nevertheless, if you're doing many remapping operations on the same grid, it may be useful to store them.

However, I think in this case there may be something preventing you from remapping or creating weights if you're planning to use CDO.

MeteoSwiss provides the grid at https://data.geo.admin.ch/browser/#/collections/ch.meteoschweiz.ogd-forecasting-icon-ch1?.language=en&.asset=asset-horizontal_constants_icon-ch1-eps.grib2 ( Horizontal Grid Parameters). Reading these files you can find variables `tlon` and `tlat` (this is really weird as usually ICON employs `clon,clat` or `vlon,vlat`). Using these coordinates you get the same plot that you show so this is already the right grid to be used.

HOWEVER, the grid file is provided in GRIB2 format, which doesn't play well with unstructured grid and cannot set the coordinates properly. That's the reason why the grid files for ICON-EU, ICON-D2,... (https://opendata.dwd.de/weather/lib/cdo/) are always provided as netcdf files.
Even when converting to netcdf the coordinates are still not properly recognized. 

```bash
> cdo griddes horizontal_constants_icon-ch1-eps.nc
#
# gridID 1
#
gridtype  = unstructured
gridsize  = 1147980
number    = 1
position  = 1
uuid      = 17643da2-5749-59b6-44d2-54a3cd6e2bc0
```

This is what a healthy grid file should look like

```
gridtype  = unstructured
gridsize  = 542040
xname     = clon
xdimname  = cell
xlongname = "longitude"
xunits    = "radian"
yname     = clat
ydimname  = cell
ylongname = "latitude"
yunits    = "radian"
vdimname  = nv
nvertex   = 3
number    = 47
position  = 0
uri       = private
xvals     = 0.0750902801597766 , ....
yvals = ...
uuid = 
```

I tried to set the coordinates manually in python but somehow there are some metadata I'm missing when writing that prevents `cdo` from properly recognizing `tlon` and `tlat` as coordinates. Which means you'll never be able to use this file directly in CDO to create the weights and/or remapping. 

One way around it would be to create the grid definition for ICON-CH1 by using a similar template and fill in the `xvals` and `yvals` as `tlon` and `tlat`, however this is a waste of space as you need to write the floats into a text file basically :) 

Once you have the input grid (either as grid description or as fixed netcdf file that CDO is recognizing) it will be possible to create the weights and remap using a target grid similar to this

```
gridtype = latlon
xsize    = 1851
ysize    = 847
xfirst   = -0.81
xinc     = 0.01
yfirst   = 42.03
yinc     = 0.01
```

I would try to choose a target grid that directly eliminated the boundaries because, as you can see in your plot, there are artifacts (as expected). 

Finally, please note that `cdo` does not have any method to remap using `RBF` (AFAIK), which is the faster and most precise algorithm to go from unstructured to regular grids. This is the method `icontools` employ, however the source code is not public. If you're planning to do a lot of remapping it may be worth it to implement a simple `RBF` algorithm into Swift instead of having to offload everything to CDO.

**Comment by patrick-zippenfenig** [comment]
Yes, my plan is to use remap with CDO and remap to a rotated lat lon grid which should be very close to the domain area. IIRC I need the triangle and edge information for the entire grid similar to DWD ICON grids: https://opendata.dwd.de/weather/lib/cdo/. Based on the triangle mesh, the remap weights can be precalculated.

This is just the most straightforward way without much development time. Based on the `tlat` and `tlon` files, I could also compute the triangles using a Delaunay algorithm, but its just unnecessary busy work ;-)

Implementing the RBF algorithm directly into Swift could be interesting. Do you know any reference implementations I could use a foundation? I cannot work with any close source code. According to [this PDF](https://www.dwd.de/DE/leistungen/opendata/help/modelle/Opendata_cdo_EN.pdf?__blob=publicationFile&v=3), the default version of CDO does not support RBF: `Note that the
barycentric interpolation and interpolation through radial basis function are not supported by default
in CDO.` Barycentric implementation would also be fairly easy to implement

> P.S. I'm not sure I get why it would be harder to use the original ICON unstructured grid in openmeteo, because most of the operations you do compare point data. Nearest neighbour lookup is much faster and easier to implement on an unstuctured grid (you only need to loop over 1 dimension).

The minimum distance search over 1 million elements takes ~10 ms. This is quite "slow". Operations on regular grids are much faster. Plotting maps will also be significantly easier. There are ways around it like `kd` search trees, but I just prefer the analytical projection equations :)

**Comment by guidocioni** [comment]
> Yes, my plan is to use remap with CDO and remap to a rotated lat lon grid which should be very close to the domain area. IIRC I need the triangle and edge information for the entire grid similar to DWD ICON grids: https://opendata.dwd.de/weather/lib/cdo/. Based on the triangle mesh, the remap weights can be precalculated.
> 
> This is just the most straightforward way without much development time. Based on the `tlat` and `tlon` files, I could also compute the triangles using a Delaunay algorithm, but its just unnecessary busy work ;-)
> 

Maybe you're right. In the past I just used the original ICON grids to remap, but I didn't consider the fact that there were additional infos inside that were needed. Honestly I only thought that CDO only needed `clon` and `clat` to do the remapping and the triangulation would be done internally, but I may be wrong... 
Anyway, yes, they need to publish the ORIGINAL grid file as for ICON-EU, ICON-D2...

> Implementing the RBF algorithm directly into Swift could be interesting. Do you know any reference implementations I could use a foundation? I cannot work with any close source code. According to [this PDF](https://www.dwd.de/DE/leistungen/opendata/help/modelle/Opendata_cdo_EN.pdf?__blob=publicationFile&v=3), the default version of CDO does not support RBF: `Note that the barycentric interpolation and interpolation through radial basis function are not supported by default in CDO.` Barycentric implementation would also be fairly easy to implement
> 

No, CDO does not have RBF, but there are many different implementations. Scipy has its own (https://github.com/scipy/scipy/blob/v1.15.3/scipy/interpolate/_rbf.py) but it's quite buggy. In the past I used many times this one https://github.com/treverhines/RBF and it seemed to work pretty well. I think they're both in C.

Otherwise you could ask for the source code of icontools (that's in Fortran as ICON) and start from there. 

> > P.S. I'm not sure I get why it would be harder to use the original ICON unstructured grid in openmeteo, because most of the operations you do compare point data. Nearest neighbour lookup is much faster and easier to implement on an unstuctured grid (you only need to loop over 1 dimension).
> 
> The minimum distance search over 1 million elements takes ~10 ms. This is quite "slow". Operations on regular grids are much faster. Plotting maps will also be significantly easier. There are ways around it like `kd` search trees, but I just prefer the analytical projection equations :)

When I mentioned NN lookup, I specifically meant KD-Trees: I thought this is what you employed already.
Actually, it's better to use `BallTree`. I used this method (https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.BallTree.html) in the past and it worked pretty well (and really fast!) to do NN lookups of hundreds of thousands of points, and I even had the Python overhead that was slowing me down :) 
Also, `BallTree` has the advantage that you can use a metric (Haversine) that already considers the spherical geometry so the lookup is really precise. Have you ever tried? I would guess nothing would be faster than a KD-Tree NN query in Swift honestly.

**Comment by patrick-zippenfenig** [comment]
I just noticed that CDO (at least the way I use it) is only doing a nearest neighbour remap. Only affects ICON global and ICON-EPS models. I honestly expected it was doing something more sophisticated. Well at least, I can easily replace this code not with a native implementation...

RBF looks really complex. Some more information in this [ICON training section 3.9.1](https://www.cosmo-model.org/content/model/documentation/core/iconTutorial_dwd_2020.pdf). Not sure If this is feasible to recreate. I am especially worried about computing those shape parameters...

> When I mentioned NN lookup, I specifically meant KD-Trees: I thought this is what you employed already.

No, I am not using any tree structures for now. All grids have analytical equations which are faster to solve than any tree lookup. `BallTree` looks interesting.

For plots, I am considering to implement the contouring directly in JavaScript to run on client side. I still have to play around with mapbox: https://github.com/open-meteo/open-data/issues/26#issuecomment-2830900526

Maybe for ICON-CH1 a simple nearest neighbour mapping also works. The mapping could also be computed without the triangle information. An open question is how well the rotated lat lon grid fits on the existing coordinates

**Comment by guidocioni** [comment]
> I just noticed that CDO (at least the way I use it) is only doing a nearest neighbour remap. Only affects ICON global and ICON-EPS models. I honestly expected it was doing something more sophisticated. Well at least, I can easily replace this code not with a native implementation...
> 

If you used `remapnn` then yes, it is nearest neighbour. When specifying that command there are slightly better methods (usually `remapcon` is the best for unstructured) 

> RBF looks really complex. Some more information in this [ICON training section 3.9.1](https://www.cosmo-model.org/content/model/documentation/core/iconTutorial_dwd_2020.pdf). Not sure If this is feasible to recreate. I am especially worried about computing those shape parameters...
> 

I gave you a challenge :P 

> > When I mentioned NN lookup, I specifically meant KD-Trees: I thought this is what you employed already.
> 
> No, I am not using any tree structures for now. All grids have analytical equations which are faster to solve than any tree lookup. `BallTree` looks interesting.
> 
> For plots, I am considering to implement the contouring directly in JavaScript to run on client side. I still have to play around with mapbox: [open-meteo/open-data#26 (comment)](https://github.com/open-meteo/open-data/issues/26#issuecomment-2830900526)
> 
> Maybe for ICON-CH1 a simple nearest neighbour mapping also works. The mapping could also be computed without the triangle information. An open question is how well the rotated lat lon grid fits on the existing coordinates

**Comment by patrick-zippenfenig** [comment]
I had a look at the RBF interpolation and I am honestly not convinced this interpolation is accurate. RBF is basically a weighted an average of the ~10 nearest points. The weight is only given by distance and the RBF function (Gauss). This is also the reason why the `epsilon` term is incredibly important. If it is too big, all details get "washed out". Too small, generates NaNs. The only benefit of RBF is aggregation to a lower resolution. Maybe I also misunderstood the implementation.

IMHO regridding on a triangle mesh using Barycentric Coordinates should be much better.

BTW: In the meantime I implemented the nearest neighbour regridding directly in Swift and removed `cdo`. Reduced the Docker images from 850MB to 200MB...

**Comment by guidocioni** [comment]
> I had a look at the RBF interpolation and I am honestly not convinced this interpolation is accurate. RBF is basically a weighted an average of the ~10 nearest points. The weight is only given by distance and the RBF function (Gauss). This is also the reason why the `epsilon` term is incredibly important. If it is too big, all details get "washed out". Too small, generates NaNs. The only benefit of RBF is aggregation to a lower resolution. Maybe I also misunderstood the implementation.
> 
> IMHO regridding on a triangle mesh using Barycentric Coordinates should be much better.
> 
> BTW: In the meantime I implemented the nearest neighbour regridding directly in Swift and removed `cdo`. Reduced the Docker images from 850MB to 200MB...

Why only 10 points? It should weight all the points with the kernel function (that's why usually it's slower than other methods). 
I never used the default Gauss kernel, thin spline works much better, and I can tell you this was one of the only methods I had in the past to interpolate sparse data onto a surface (which would otherwise not be approximated correctly).

To be honest, the only way to test what is best would ONLY be to perform the 2 different interpolations and evaluate the error with a point to point difference metric. Otherwise, the "feeling" is never right :)
I never had time to do a proper comparison, so I trusted the fact that this is the default method used when creating initial and boundary conditions on the ICON grid: AFAIK, it's the best (in terms of accuracy and speed) method to go from every grid to ICON grid, or from an ICON grid to another ICON grid. I don't know if that's the case when you're interpolating to a regular one.

But then again, I know nothing. If you want to know more about it maybe you can try to contact Florian Prill at DWD: he was the one who implemented this method in the ICON tools (the set of binaries used to create initial and boundary conditions for ICON).

**Comment by patrick-zippenfenig** [comment]
Form the ICON training PDF. For all points the runtime of O(N^2) would also be very expensive...

> For cell based variables a 2D horizontal gradient is reconstructed at the parent
cell center by first computing edge-normal gradients at edge midpoints, followed by a
9-point reconstruction of the 2D gradient at the cell center based on radial basis functions
(RBF, Narcovich and Ward (1994))

I will have to do more research. It also gets more complex because some fields like wind are defined on the edge of the triangle. Others like temperature on the triangle centre (or outer point?). It would be so much easier with just a simple stereographic projection grid....

**Comment by guidocioni** [comment]
> Form the ICON training PDF. For all points the runtime of O(N^2) would also be very expensive...
> 
> > For cell based variables a 2D horizontal gradient is reconstructed at the parent
> > cell center by first computing edge-normal gradients at edge midpoints, followed by a
> > 9-point reconstruction of the 2D gradient at the cell center based on radial basis functions
> > (RBF, Narcovich and Ward (1994))
> 
> I will have to do more research. It also gets more complex because some fields like wind are defined on the edge of the triangle. Others like temperature on the triangle centre (or outer point?). It would be so much easier with just a simple stereographic projection grid....

Yeah it's a pain to postprocess edges, vertices...
There's no ICON-CH1 product already remapped on regular? 
Again, I still think the best option is storing the data in the original grid if the goal is to make a NN lookup. Remapping only makes sense if you want to somehow produce rasters and have less compatibility issues...but it's up to you :)

**Comment by patrick-zippenfenig** [comment]
> There's no ICON-CH1 product already remapped on regular?

Not that I am aware of. The new ICON-D2-RUC (#1368) is also available native only. So far I am only doing NN for ICON global and all EPS domains.

> Again, I still think the best option is storing the data in the original grid if the goal is to make a NN lookup.

It would work, but any application "down the line" needs to work with it. Especially for [maps](https://open-meteo.github.io/omaps/#2.96/48.91/13.19), it's not ideal. And the future OM file Python reader would also need to work with native ICON grids...

**Comment by patrick-zippenfenig** [comment]
Well... spend the last 6 hours playing around with interpolations. All experiments use ICON-D2 on a regular lat lon grid. I haven't looked at triangle grids for ICON-CH1 yet

RBF looks smooth with 4x4 points. However, the epsilon parameter needs to be selected very carefully, otherwise it is immediately unsteady (haven't successfully implemented Cholesky decomposition). It certainly smoothes values.

![Image](https://github.com/user-attachments/assets/21a479bf-90f5-475c-8d73-44e701e5e4b5)

I prefer a 2D Hermite cubic cardinal considering 4x4 points. Of course this only works 2D grids and not with triangle meshes. Currently live [here](https://open-meteo.github.io/omaps/).

![Image](https://github.com/user-attachments/assets/bab52896-c01b-421e-9435-1f3bbd8a712f)

Bilinear Is also not too bad:

![Image](https://github.com/user-attachments/assets/ffcdb7e2-d26a-4eb7-bd25-47621bbdac28)

Still undecided what to do either ICON-CH1.... I will check how the well the grid cells of a stereographic projection align. Maybe a simple NN is fine.

**Comment by guidocioni** [comment]
I think you're right: if your target grid is not icosahedral RBF is not the best choice. 
I mostly used RBF when the target grid (and sometimes the original grid) were icosahedral, or when interpolating station data.

In other cases more 'stable' methods may be better.

Still, remember it always depends on the particular implementation.
We use RBF for station interpolation in Meteonetwork in production and never had singularities or weird stuff happening. Of course outside of the domain the function is not stable, but that's where you don't have informationl...

**Comment by patrick-zippenfenig** [comment]
Current state: Download and reprojecting to rotated lat lon are working. Now waiting for feedback from MeteoSwiss regarding https://github.com/MeteoSwiss/meteodata-lab/issues/103 https://github.com/open-meteo/open-meteo/issues/1286
